gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java

```diff
@@ -104,52 +113,82 @@ private void addLoadingElement(CompletionResultSet sortedResultSet) {
+                .collect(Collectors.toList());
+        addToResults(sortedResultSet, allPackages, dependencyInfo);
     }
 
-    private List<PackageInRepo> collectGroupAndDeps(
-            Project project, DependencyGroup group, DependencyName dependencyName) {
+    private List<PackageInRepo> collectPackageInRepo(Project project, DependencyInfo dependencyInfo) {
 
-        String dependencyNamePrefix = dependencyName.name().replace("*", "");
+        String dependencyNamePrefix = dependencyInfo.dependencyName().name().replace("*", "");
         return StreamEx.of(RepositoryLoader.loadRepositories(project))
-                .flatMap(url -> StreamEx.of(
-                                groupPartOrPackageNameExplorer.getCancelableGroupPartOrPackageName(group, url))
+                .flatMap(url -> StreamEx.of(groupPartOrPackageNameExplorer.getCancelableGroupPartOrPackageName(
+                                dependencyInfo.group(), url))
                         .filter(pkgName -> pkgName.name().startsWith(dependencyNamePrefix))
                         .map(pkgName -> new SimpleEntry<>(url, pkgName)))
                 .map(entry -> {
                     RepositoryUrl url = entry.getKey();
                     GroupPartOrPackageName pkgName = entry.getValue();
                     DependencyName depName = DependencyName.of(pkgName.name());
-                    return new PackageInRepo(group, depName, url);
+                    return new PackageInRepo(dependencyInfo.group(), depName, url);
                 })
                 .toList();
     }
 
-    private void addToResults(CompletionResultSet resultSet, List<PackageInRepo> groupAndDeps) {
-        Map<DependencyVersion, AtomicInteger> versionCounts = StreamEx.of(groupAndDeps)
-                .flatMap(groupAndDep ->
-                        versionExplorer
-                                .getVersions(groupAndDep, CompletionRefreshUtil.refreshOnceSupplier()::get)
-                                .stream())
-                .collect(Collectors.toConcurrentMap(
-                        Function.identity(), v -> new AtomicInteger(1), (existingCount, newCount) -> {
-                            existingCount.addAndGet(newCount.get());
-                            return existingCount;
-                        }));
+    private void addToResults(CompletionResultSet resultSet, List<PackageInRepo> packageInRepo, DependencyInfo key) {
+
+        Set<VersionResults> versionResults =
+                packageInRepo.stream().map(versionExplorer::getVersions).collect(Collectors.toSet());
+
+        List<CompletableFuture<?>> pendingFutures = versionResults.stream()
+                .filter(results -> results.versions().isEmpty())
+                .map(VersionResults::future)
+                .filter(future -> !future.isDone())
+                .collect(Collectors.toList());
+
+        if (!pendingFutures.isEmpty()) {
+            CompletableFuture.anyOf(pendingFutures.toArray(new CompletableFuture[0]))
```

Comment:
This should trigger a refresh when any one of the futures completes which should schedule less refreshes than just doing a refresh for each future I think